### ISSUE TITLE
Device (node, mqtt) disconnects after couple of days

### ISSUE THREAD

**Issue by tameraw** [Azure/azure-iot-sdk-node#12]
_From @bmrpatel on November 22, 2016 12:53_

**Platform: Linux, Nodejs 7.1, SDK v 1.1 MQTT**

Device getting disconnected is a very serious issue and can't deploy it to prod unless resolved. Only way to resolve this is to restart the device.

_**Scenario:**_ Device is connected and not getting any commands or methods from cloud for a day or two.
 It gets disconnected after 2nd day almost. `disconnect` event is not fired as well. 
**Question** Do I have to some some sort of heartbeat to let server know that device is up and running?

_Copied from original issue: Azure/azure-iot-sdks#980_

**Comment by tameraw** [comment]
_From @bmrpatel on November 22, 2016 12:56_

may be related to #981

**Comment by tameraw** [comment]
_From @bmrpatel on November 29, 2016 22:3_

I had two test devices connected for 3 days and today both of them are disconnected with the following debug log: Please note that both devices got disconnected at the same time. Hope this will help investigate this issue.

**device1**
```
Tue, 29 Nov 2016 17:44:52 GMT mqtt-common topic publish: devices/000019/messages/events/
Tue, 29 Nov 2016 17:44:52 GMT mqtt-common topic subscribe: devices/000019/messages/devicebound/#
Tue, 29 Nov 2016 17:44:52 GMT mqtt-common username: myhubendpoint.azure-devices.net/000019/DeviceClientType=azure-iot-device%2F1.1.0&api-version=2016-11-14
Tue, 29 Nov 2016 17:44:52 GMT mqtt-common uri:      mqtts://myhubendpoint.azure-devices.net
Tue, 29 Nov 2016 17:44:52 GMT mqtt-common Device is connected
Tue, 29 Nov 2016 17:44:52 GMT mqtt-common CONNACK: {"cmd":"connack","retain":false,"qos":0,"dup":false,"length":2,"topic":null,"payload":null,"sessionPresent":true,"returnCode":0}
Tue, 29 Nov 2016 18:06:44 GMT mqtt-common Device connection to the server has been closed

```

**device 2**
```
Tue, 29 Nov 2016 16:13:54 GMT mqtt-common uri:      mqtts://myhubendpoint.azure-devices.net
Tue, 29 Nov 2016 16:13:55 GMT mqtt-common Device is connected
Tue, 29 Nov 2016 16:13:55 GMT mqtt-common CONNACK: {"cmd":"connack","retain":false,"qos":0,"dup":false,"length":2,"topic":null,"payload":null,"sessionPresent":true,"returnCode":0}
Tue, 29 Nov 2016 16:58:55 GMT mqtt-common topic publish: devices/001262/messages/events/
Tue, 29 Nov 2016 16:58:55 GMT mqtt-common topic subscribe: devices/001262/messages/devicebound/#
Tue, 29 Nov 2016 16:58:55 GMT mqtt-common username: myhubendpoint.azure-devices.net/001262/DeviceClientType=azure-iot-device%2F1.1.0&api-version=2016-11-14
Tue, 29 Nov 2016 16:58:55 GMT mqtt-common uri:      mqtts://myhubendpoint.azure-devices.net
Tue, 29 Nov 2016 16:58:55 GMT mqtt-common Device is connected
Tue, 29 Nov 2016 16:58:55 GMT mqtt-common CONNACK: {"cmd":"connack","retain":false,"qos":0,"dup":false,"length":2,"topic":null,"payload":null,"sessionPresent":true,"returnCode":0}
Tue, 29 Nov 2016 17:43:56 GMT mqtt-common topic publish: devices/001262/messages/events/
Tue, 29 Nov 2016 17:43:56 GMT mqtt-common topic subscribe: devices/001262/messages/devicebound/#
Tue, 29 Nov 2016 17:43:56 GMT mqtt-common username: myhubendpoint.azure-devices.net/001262/DeviceClientType=azure-iot-device%2F1.1.0&api-version=2016-11-14
Tue, 29 Nov 2016 17:43:56 GMT mqtt-common uri:      mqtts://myhubendpoint.azure-devices.net
Tue, 29 Nov 2016 17:43:56 GMT mqtt-common Device is connected
Tue, 29 Nov 2016 17:43:56 GMT mqtt-common CONNACK: {"cmd":"connack","retain":false,"qos":0,"dup":false,"length":2,"topic":null,"payload":null,"sessionPresent":true,"returnCode":0}
Tue, 29 Nov 2016 18:06:47 GMT mqtt-common Device connection to the server has been closed
```

**Comment by tameraw** [comment]
_From @sergaz-msft on January 12, 2017 12:2_

Hi @bmrpatel , thank you for letting us know about this. I will try to reproduce the same on my Environment. Can you please clarify the following?

"Scenario: Device is connected and not getting any commands or methods from cloud for a day or two.
It gets disconnected after 2nd day almost. disconnect event is not fired as well."

This means that device is not receiving Cloud-to-device messages and not sending device-to-Cloud messages during 2 days correct?

Thanks

**Comment by tameraw** [comment]
_From @bmrpatel on January 12, 2017 15:50_

Thanks for looking into this issue...

As far as I can find out, it has been disconnecting with the debug message: 
`mqtt-common Device connection to the server has been closed`

This happens when device is busy processing some other long task. So my comments about 2 days may be irrelevant. It may happened to be after 2 days at that time, but it can happen right away if I start he lengthy task.


**Off topic question:** I have also notices (very rarely) that **device explorer** tool is not showing all of my devices on the Management tab. I also can't connect to cloud from device at that time. It seems as if IOT Hub service might be down at that time. Is it possible?

**Comment by bmrpatel** [comment]
@tameraw @sergaz-msft The easiest way to reproduce this is to set a break point after connection is made. Wait for 30-40 seconds and then continue, you will see the below message and get disconnected.

`mqtt-common Device connection to the server has been closed`

**Comment by bmrpatel** [comment]
@sergaz-msft this happened again last night:

**Device 1**
```
Jan 14 05:58:22 arm index.js[3357]: Sat, 14 Jan 2017 05:58:22 GMT mqtt-common topic publish: devices/0001262/messages/events/
Jan 14 05:58:22 arm index.js[3357]: Sat, 14 Jan 2017 05:58:22 GMT mqtt-common topic subscribe: devices/0001262/messages/devicebound/#
Jan 14 05:58:22 arm index.js[3357]: Sat, 14 Jan 2017 05:58:22 GMT mqtt-common username: {name}.azure-devices.net/0001262/DeviceClientType=azure-iot-device%2F1.1.2&api-version=2016-11-14
Jan 14 05:58:22 arm index.js[3357]: Sat, 14 Jan 2017 05:58:22 GMT mqtt-common uri:      mqtts://pos-intg.azure-devices.net
Jan 14 05:58:23 arm index.js[3357]: Sat, 14 Jan 2017 05:58:23 GMT mqtt-common Device is connected
Jan 14 05:58:23 arm index.js[3357]: Sat, 14 Jan 2017 05:58:23 GMT mqtt-common CONNACK: {"cmd":"connack","retain":false,"qos":0,"dup":false,"length":2,"topic":null,"payload":null,"sessionPresent":true,"returnCode":0}
Jan 14 06:19:34 arm index.js[3357]: Sat, 14 Jan 2017 06:19:34 GMT mqtt-common Device connection to the server has been closed
```
**Device 2**
```
Jan 14 05:35:03 pos000019 index.js[12366]: Sat, 14 Jan 2017 05:35:03 GMT mqtt-common uri:      mqtts://pos-intg.azure-devices.net
Jan 14 05:35:04 pos000019 index.js[12366]: Sat, 14 Jan 2017 05:35:04 GMT mqtt-common Device is connected
Jan 14 05:35:04 pos000019 index.js[12366]: Sat, 14 Jan 2017 05:35:04 GMT mqtt-common CONNACK: {"cmd":"connack","retain":false,"qos":0,"dup":false,"length":2,"topic":null,"payload":null,"sessionPresent":true,"returnCode":0}
Jan 14 06:19:37 pos000019 index.js[12366]: Sat, 14 Jan 2017 06:19:37 GMT mqtt-common Device connection to the server has been closed
```

As you can see both devices got disconnected at about same time, Let me know if you need more info?

Thanks

**Comment by sergaz-msft** [comment]
Hi @bmrpatel , I was able to reproduce the issue on my side. I guess that you are using [simple_sample_device.js](https://github.com/Azure/azure-iot-sdk-node/blob/master/device/samples/simple_sample_device.js) and on that sample we instruct the device to reconnect if it gets disconnected 
```node
 client.on('disconnect', function () {
      clearInterval(sendInterval);
     client.removeAllListeners();
      client.open(connectCallback);
    });
```
As you correctly identified, the device doesn't reconnect because it never enters this function.

I found the solution for that (will do a pull request after checking that it doesn't break anything else).

We need to change the following in this node module [azure-iot-device-mqtt\mqtt.js#Line57](https://github.com/Azure/azure-iot-sdk-node/blob/master/device/transport/mqtt/lib/mqtt.js#L57)

```node
this._mqtt.client.on('close', function (err) {
```
After that the device will reconnect after long tasks.
Please let me know if this also solves the issue #11. If not I will look at it after.

Regarding the **off topic question**, when you see the issue can you check with iothub-explorer tool to [export all devices](https://github.com/Azure/iothub-explorer/blob/master/iothub-explorer-export-devices.js) from your IoT Hub and see if all devices are there? If you continue to see the issue please open a new ticket under [Device Explorer](https://github.com/Azure/azure-iot-sdk-csharp/tree/master/tools/DeviceExplorer).

Cheers.

**Comment by bmrpatel** [comment]
Hello @sergaz-msft, This is great! please let me know and I will update my package or code manually to test this change. Actually, can I just make that change on line 57 and test myself locally? 

I will let you know if this will resolve the other issue #11.

Thanks

**Comment by sergaz-msft** [comment]
Hi @bmrpatel , if your environment is for tests sure please give it a try and change line 57 :)!

**Comment by BertKleewein** [comment]
I just committed a fix for this to the master branch.  It should be present in our next release.

**Comment by pierreca** [comment]
Adding the link to the fix here: https://github.com/Azure/azure-iot-sdk-node/commit/a46c4a1de0008e9c61f0b86bce8d54965b438914
and marking the issue as fix-checked-in. will be closed when it's released.

**Comment by bmrpatel** [comment]
Thanks It seems to be working now.

**Comment by pierreca** [comment]
The fix has been released to npm with packages version 1.1.4:
- https://www.npmjs.com/package/azure-iot-device-mqtt
- https://www.npmjs.com/package/azure-iot-device
